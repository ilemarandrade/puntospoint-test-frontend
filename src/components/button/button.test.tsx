import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Button from '.';
import AddIcon from '@mui/icons-material/Add';

describe('Button Component', () => {
  test('renders the button with default variant', () => {
    render(<Button>Label</Button>);
    const buttonElement = screen.getByRole('button', { name: /label/i });
    expect(buttonElement).toBeInTheDocument();
  });

  test('does not trigger onClick when disabled', () => {
    const handleClick = jest.fn();
    render(
      <Button onClick={handleClick} disabled>
        Label
      </Button>
    );

    const buttonElement = screen.getByRole('button', { name: /label/i });

    expect(buttonElement).toBeDisabled();

    fireEvent.click(buttonElement);

    expect(handleClick).not.toHaveBeenCalled();
  });

  test('does not trigger onClick and hides label when loading', () => {
    const handleClick = jest.fn();
    render(
      <Button onClick={handleClick} loading>
        Label
      </Button>
    );

    const buttonElement = screen.getByRole('button', { name: /label/i });
    const loadingElement = screen.getByRole('progressbar');

    expect(buttonElement).toBeDisabled();

    expect(loadingElement).toBeVisible();

    fireEvent.click(buttonElement);

    expect(handleClick).not.toHaveBeenCalled();
  });

  test('renders the button with different variants', () => {
    const variants: ['contained', 'outlined', 'text'] = [
      'contained',
      'outlined',
      'text',
    ];

    variants.forEach((variant) => {
      const { container } = render(
        <Button variant={variant}>{variant}</Button>
      );

      expect(container).toMatchSnapshot();

      const buttonElement = screen.getByRole('button', { name: variant });
      expect(buttonElement).toHaveClass(`MuiButton-${variant}`);
    });
  });

  test('displays the correct icon', () => {
    const { container } = render(<Button startIcon={<AddIcon />} />);

    expect(container).toMatchSnapshot();
    const iconElement = screen.getByTestId('AddIcon');
    expect(iconElement).toBeInTheDocument();
  });

  test('handles onClick event', () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Label</Button>);
    const buttonElement = screen.getByRole('button', { name: /label/i });
    fireEvent.click(buttonElement);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test('renders button with href', () => {
    render(<Button href="https://example.com">Go to Example</Button>);
    const buttonElement = screen.getByRole('link', { name: /go to example/i });
    expect(buttonElement).toHaveAttribute('href', 'https://example.com');
  });

  test('renders button with custom props', () => {
    render(<Button data-testid="custom-button">Custom Button</Button>);
    const buttonElement = screen.getByTestId('custom-button');
    expect(buttonElement).toBeInTheDocument();
  });

  test('renders button label correctly', () => {
    const label = 'Submit';
    render(<Button>{label}</Button>);
    const buttonElement = screen.getByRole('button', { name: /submit/i });
    expect(buttonElement).toHaveTextContent(label);
  });
});
