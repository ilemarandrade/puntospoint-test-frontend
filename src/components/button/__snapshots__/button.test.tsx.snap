// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component displays the correct icon 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1sh91j5-MuiButton-startIcon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`Button Component renders the button with different variants 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    contained
  </button>
</div>
`;

exports[`Button Component renders the button with different variants 2`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1blvszr-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    outlined
  </button>
</div>
`;

exports[`Button Component renders the button with different variants 3`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1uent87-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    text
  </button>
</div>
`;
